src/loop/parser.ml

```diff
@@ -190,70 +192,131 @@ module Make(State : State.S) = struct
-              Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
-          in
-          let file = { file with lang = Some lang; } in
-          st, file, (Gen.singleton s')
-      with
-      | Logic.Extension_not_found ext ->
-        State.error st extension_not_found ext, file, Gen.empty
+    match source with
+    | `Raw (filename, contents) ->
+      begin match file.mode with
+      | None
+      | Some `Incremental ->
+        let lang, file_loc, gen, cl = Logic.parse_input
+            ~language:lang (`Raw (filename, lang, contents)) in
+        let file = { file with loc = file_loc; lang = Some lang; } in
+        st, file, gen_finally gen cl
+      | Some `Full ->
+        let lang, file_loc, l = Logic.parse_raw_lazy ~language:lang
+          ~filename contents
+        in
+        let file = { file with loc = file_loc; lang = Some lang; } in
+        (* We are called from inside a [gen'] so the returned generator is
+        always evaluated immediately, there is no need to add laziness *)
+        st, file, Gen.of_list (Lazy.force l)
+      end
+    | `File f ->
+      let s = Dolmen.Std.Statement.include_ f [] in
+      (* Formats Dimacs and Tptp are descriptive and lack the emission
+          of formal solve/prove instructions, so we need to add them. *)
+      let s' =
+        match lang with
+        | Logic.Zf
+        | Logic.ICNF
+        | Logic.Smtlib2 _
+        | Logic.Alt_ergo -> s
+        | Logic.Dimacs
+        | Logic.Tptp _ ->
+          Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
+      in
+      let file = { file with lang = Some lang; } in
+      st, file, (Gen.singleton s')
+
+  (* This is a ['a Gen.t] with a threaded state. *)
+  type 'a gen' = {g: State.t -> State.t * 'a option} [@@unboxed]
+
+  let make g = {g}
+
+  let run {g} = g
+
+  let flat_map (f : 'a -> 'b gen') (g : 'a gen') : 'b gen' =
+    (* [cur_f] holds the last generator returned by a call to [f] *)
+    let cur_f = ref None in
+    let rec aux st =
+      match !cur_f with
+      | None ->
+        (* [cur_f] is empty, get a new one and try again *)
+        let st, x = run g st in
+        begin match x with
+        | None -> st, None
+        | Some x ->
+          cur_f := Some (f x);
+          aux st
+        end
+      | Some c ->
+        (* We got a generator, use it or refill it *)
+        let st, c' = run c st in
+        match c' with
+        | Some _ -> st, c'
+        | None ->
+          cur_f := None;
+          aux st
     in
-    let st = set_logic_file st file in
-    (* Wrap the resulting parser *)
-    st, wrap_parser ~file (Gen.append (Gen.of_list prelude) g)
+    make aux
+
+  (* [map_st_gen f g] calls [f] on all the values produced by [g] in order,
+     threading the state.
+
+     [f] is *not* a generator but a function in the state monad. *)
+  let map_st_gen f (g : _ Gen.t) : _ gen' =
+    make @@ fun st ->
+    begin match g () with
+    | None -> st, None
+    | Some x -> f x st
```

Comment:
Is there a reason to put `x` first here ? Intuitively I would rather have `st` first, and then `x`

Reply:
I tried to keep the `state -> state * 'a` intact because I think of it as a single type constructor; I can flip the arguments around.